x/nft/keeper/keeper.go

```diff
@@ -0,0 +1,152 @@
+
+	"github.com/crypto-org-chain/chain-main/v2/x/nft/types"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	storeKey sdk.StoreKey // Unexposed key to access store from sdk.Context
+	cdc      codec.Marshaler
+}
+
+// NewKeeper creates a new instance of the NFT Keeper
+func NewKeeper(cdc codec.Marshaler, storeKey sdk.StoreKey) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("chainmain/%s", types.ModuleName))
+}
+
+// IssueDenom issues a denom according to the given params
+func (k Keeper) IssueDenom(ctx sdk.Context,
+	id, name, schema string,
+	creator sdk.AccAddress) error {
+	return k.SetDenom(ctx, types.NewDenom(id, name, schema, creator))
+}
+
+// MintNFT mints an NFT and manages the NFT's existence within Collections and Owners
+func (k Keeper) MintNFT(
+	ctx sdk.Context, denomID, tokenID, tokenNm,
+	tokenURI, tokenData string, owner sdk.AccAddress,
+) error {
+	if !k.HasDenomID(ctx, denomID) {
+		return sdkerrors.Wrapf(types.ErrInvalidDenom, "denom ID %s not exists", denomID)
+	}
+
+	if k.HasNFT(ctx, denomID, tokenID) {
+		return sdkerrors.Wrapf(types.ErrNFTAlreadyExists, "NFT %s already exists in collection %s", tokenID, denomID)
+	}
+
+	k.setNFT(
+		ctx, denomID,
+		types.NewBaseNFT(
+			tokenID,
+			tokenNm,
+			owner,
+			tokenURI,
+			tokenData,
+		),
+	)
+	k.setOwner(ctx, denomID, tokenID, owner)
+	k.increaseSupply(ctx, denomID)
+
+	return nil
+}
+
+// EditNFT updates an already existing NFT
+func (k Keeper) EditNFT(
+	ctx sdk.Context, denomID, tokenID, tokenNm,
+	tokenURI, tokenData string, owner sdk.AccAddress,
+) error {
+	if !k.HasDenomID(ctx, denomID) {
+		return sdkerrors.Wrapf(types.ErrInvalidDenom, "denom ID %s not exists", denomID)
+	}
+
+	nft, err := k.Authorize(ctx, denomID, tokenID, owner)
+	if err != nil {
+		return err
+	}
+
+	if types.Modified(tokenNm) {
+		nft.Name = tokenNm
+	}
+
+	if types.Modified(tokenURI) {
+		nft.URI = tokenURI
+	}
+
+	if types.Modified(tokenData) {
+		nft.Data = tokenData
+	}
+
+	k.setNFT(ctx, denomID, nft)
+
+	return nil
+}
+
+// TransferOwner transfers the ownership of the given NFT to the new owner
+func (k Keeper) TransferOwner(
+	ctx sdk.Context, denomID, tokenID, tokenNm, tokenURI,
+	tokenData string, srcOwner, dstOwner sdk.AccAddress,
+) error {
+	if !k.HasDenomID(ctx, denomID) {
+		return sdkerrors.Wrapf(types.ErrInvalidDenom, "denom ID %s not exists", denomID)
+	}
+
+	nft, err := k.Authorize(ctx, denomID, tokenID, srcOwner)
+	if err != nil {
+		return err
+	}
+
+	nft.Owner = dstOwner.String()
+
+	if types.Modified(tokenNm) {
+		nft.Name = tokenNm
+	}
+
+	if types.Modified(tokenURI) {
+		nft.URI = tokenURI
+	}
+
+	if types.Modified(tokenData) {
+		nft.Data = tokenData
+	}
```

Comment:
ummm... so each transfer can potentially change the NFT content?

Reply:
Yeah. This is weird. Ideally, `TransferNFT` should only take `nftId` and `toAddress`? And it should only make the transfer if the message is signed by the owner of the NFT?